java/java-impl/src/com/intellij/codeInsight/FunctionalInterfaceSuggester.java

```diff
@@ -260,22 +253,28 @@ else if (expression instanceof PsiMethodReferenceExpression) {
 
                 for (int i = 0; i < targetMethodParameters.length; i++) {
                   left[i + offset] = parameters[i + offset].getType();
-                  right[i + offset] = partialSubstitutor.substitute(targetMethodParameters[i].getType());
+                  right[i + offset] = targetMethodParameters[i].getType();
                 }
 
+                final PsiSubstitutor partialSubstitutor = getPartialSubstitutor(referenceExpression, method);
+
                 left[parameters.length] = method.isConstructor() ? qualifierType : partialSubstitutor.substitute(method.getReturnType());
                 right[parameters.length] = returnType;
 
-                final PsiSubstitutor substitutor = PsiResolveHelper.SERVICE.getInstance(project)
-                  .inferTypeArguments(interface2Consider.getTypeParameters(), left, right, PsiUtil.getLanguageLevel(expression));
-
-                PsiType type = JavaPsiFacade.getElementFactory(project).createType(interface2Consider, substitutor);
+                final Optional<PsiType> type = getAcceptableType(interface2Consider, expression, left, right, partialSubstitutor);
 
```

Comment:
return an optional to add if present in both branches to the same collection, looks like an overcomplicating to me, just pass the collection and fill it when applicable

Reply:
Thank you for your review.
I looked inside of `Optional#ifPresent` and it does the same thing: it checks if the value is not null and is so the `Consumer` is invoked. It seemed too me as a good idea to let the standard library do the job. But you might have different standards and conventions in your team, so I will get rid of `Optional` and bring back the original code.

Reply:
I removed `Optional` here